core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -57,6 +58,37 @@ class KafkaZkClient(zooKeeperClient: ZooKeeperClient, isSecure: Boolean) extends
     createResponse.path
   }
 
+  private var brokerInfo: Array[Byte] = null;
+  private var brokerId: Int = -1;
+
+  def registerBrokerInZk(id: Int,
+                         host: String,
+                         port: Int,
+                         advertisedEndpoints: Seq[EndPoint],
+                         jmxPort: Int,
+                         rack: Option[String],
+                         apiVersion: ApiVersion) {
+    val brokerIdPath = BrokerIdsPath + "/" + id
+    // see method documentation for reason why we do this
+    val version = if (apiVersion >= KAFKA_0_10_0_IV1) 4 else 2
+    val json = Broker.toJson(version, id, host, port, advertisedEndpoints, jmxPort, rack)
+    brokerInfo = json.getBytes()
+    brokerId = id
+    checkedEphemeralCreate(brokerIdPath, brokerInfo)
+
+    info("Registered broker %d at path %s with addresses: %s".format(id, brokerIdPath, advertisedEndpoints.mkString(",")))
+  }
```

Comment:
We probably can decouple the creation of brokerInfo from the registration in ZK. Perhaps we can add a method in ZkData.scala to create brokerInfo and have a method registerBrokerInZk(brokerInfo) to take in brokerInfo. In KafkaServer.startup(), we first create brokerInfo and call registerBrokerInZk(brokerInfo) to do the initial broker registration. We then pass in brokerInfo to KafkaController so that it can be used to re-register the broker when the ZK session expires.

Reply:
Makes sense, passing BrokerInfo explicitly to Controller sounds better than keeping private variables in KafkaZkClient to recreate (that can be called by anyone and not just Controller)